README.md

## bigFastlm [![version](http://www.r-pkg.org/badges/version/bigFastlm)](https://cran.r-project.org/package=bigFastlm)
A reimplementation of the fastLm functionality of RcppEigen for
big.matrix objects for fast out-of-memory linear model fitting


### Build Status
|  OS                   | Build           |
|-----------------------|-----------------|
| Linux x86_64          | [![Build Status](https://travis-ci.org/jaredhuling/bigFastlm.svg?branch=master)](https://travis-ci.org/jaredhuling/bigFastlm)      | 
| Windows x86_64        | [![Appveyor Build Status](https://ci.appveyor.com/api/projects/status/github/jaredhuling/bigFastlm?branch=master&svg=true)](https://ci.appveyor.com/project/jaredhuling/bigFastlm)     |



### Installation

Install using the **devtools** package (RcppEigen and bigmemory must be installed first as well):

```r
devtools::install_github("jaredhuling/bigFastlm")
```


### Usage

```r
library(bigFastlm)
library(bigmemory)

nrows <- 50000
ncols <- 50
bkFile <- "bigmat.bk"
descFile <- "bigmatk.desc"
bigmat <- filebacked.big.matrix(nrow=nrows, ncol=ncols, type="double",
                                backingfile=bkFile, backingpath=".",
                                descriptorfile=descFile,
                                dimnames=c(NULL,NULL))

set.seed(123)
for (i in 1:ncols) bigmat[,i] = rnorm(nrows)*i

y <- rnorm(nrows) + bigmat[,1]

system.time(lmr1 <- bigLm(bigmat, y))

summary(lmr1)

predictions <- predict(lmr1, newdata = bigmat)
```


NEWS.md
# bigFastLm 0.0.2

* Fix returning of `colnames` as `NA` for coefficient names (#1).



R/bigmemory.R


big.crossprod <- function(x)
{
  if (!is.big.matrix(x))
  {
    stop("object must be a big.matrix object")
  }
  .Call("crossprod_big", x@address)
}

big.colSums <- function(x)
{
  if (!is.big.matrix(x))
  {
    stop("object must be a big.matrix object")
  }
  .Call("colsums_big", x@address)
}

big.colMax <- function(x)
{
  if (!is.big.matrix(x))
  {
    stop("object must be a big.matrix object")
  }
  .Call("colmax_big", x@address)
}

big.colMin <- function(x)
{
  if (!is.big.matrix(x))
  {
    stop("object must be a big.matrix object")
  }
  .Call("colmin_big", x@address)
}


#' big.matrix prod
#'
#' @param x big.matrix
#' @param y numeric vector
#' @docType methods
#' @importFrom methods new
#' @rdname extract-methods
setMethod("%*%",signature(x="big.matrix", y="vector"),
          function(x, y)
          {
            if(dim(x)[2] != length(y)) stop("non-conformant matrices")
            return( .Call("prod_vec_big", x@address, y) )
          },
          valueClass="vector"
)

#' big.matrix prod
#'
#' @docType methods
#' @rdname extract-methods
setMethod("%*%",signature(x="vector", y="big.matrix"),
          function(x, y)
          {
            if(dim(y)[2] != length(x)) stop("non-conformant matrices")
            return( .Call("prod_vec_big_right", x, y@address) )
          },
          valueClass="vector"
)




src/big_fast_lm.h
// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; tab-width: 8 -*-
//
// fastLm.h: Rcpp/Eigen example of a simple lm() alternative
//
// Copyright (C) 2011 - 2015  Douglas Bates, Dirk Eddelbuettel and Romain Francois
//
// This file is part of RcppEigen.
//
// RcppEigen is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// RcppEigen is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// in file.path(R.home("share"), "licenses").  If not, see
// <http://www.gnu.org/licenses/>.
#ifndef bigFastlm_BIGFASTLM_H
#define bigFastlm_BIGFASTLM_H

#include <bigmemory/MatrixAccessor.hpp>
#include <bigmemory/BigMatrix.h>
#include <Rcpp.h>
#include <RcppEigen.h>

namespace lmsol {
  using Eigen::ArrayXd;
  using Eigen::ColPivHouseholderQR;
  using Eigen::ComputeThinU;
  using Eigen::ComputeThinV;
  using Eigen::HouseholderQR;
  using Eigen::JacobiSVD;
  using Eigen::LDLT;
  using Eigen::LLT;
  using Eigen::Lower;
  using Eigen::Map;
  using Eigen::MatrixXd;
  using Eigen::SelfAdjointEigenSolver;
  using Eigen::SelfAdjointView;
  using Eigen::TriangularView;
  using Eigen::VectorXd;
  using Eigen::Upper;

  typedef MatrixXd::Index                                 Index;
  typedef MatrixXd::Scalar                                Scalar;
  typedef MatrixXd::RealScalar                            RealScalar;
  typedef ColPivHouseholderQR<MatrixXd>::PermutationType  Permutation;
  typedef Permutation::IndicesType                        Indices;

  class lm {
  protected:
    Map<MatrixXd> m_X;	/**< model matrix */
    Map<VectorXd> m_y;	/**< response vector */
    Index         m_n;	/**< number of rows of X */
    Index         m_p;	/**< number of columns of X */
    VectorXd      m_coef;	/**< coefficient vector */
    int           m_r;	/**< computed rank or NA_INTEGER */
    VectorXd      m_fitted;	/**< vector of fitted values */
    VectorXd      m_se;	/**< standard errors  */
    RealScalar    m_prescribedThreshold; /**< user specified tolerance */
    bool          m_usePrescribedThreshold;
  public:
    lm(const Map<MatrixXd>&, const Map<VectorXd>&);

    ArrayXd                Dplus(const ArrayXd& D);
    MatrixXd                 I_p() const {return MatrixXd::Identity(m_p, m_p);}
    MatrixXd                 XtX() const;

    // setThreshold and threshold are based on ColPivHouseholderQR methods
    RealScalar         threshold() const;
    const VectorXd&           se() const {return m_se;}
    const VectorXd&         coef() const {return m_coef;}
    const VectorXd&       fitted() const {return m_fitted;}
    int                     rank() const {return m_r;}
    lm&             setThreshold(const RealScalar&);
  };

  class ColPivQR : public lm {
  public:
    ColPivQR(const Map<MatrixXd>&, const Map<VectorXd>&);
  };

  class Llt : public lm {
  public:
    Llt(const Map<MatrixXd>&, const Map<VectorXd>&);
  };

  class Ldlt : public lm {
  public:
    Ldlt(const Map<MatrixXd>&, const Map<VectorXd>&);
  };

  class QR : public lm {
  public:
    QR(const Map<MatrixXd>&, const Map<VectorXd>&);
  };

  class GESDD : public lm {
  public:
    GESDD(const Map<MatrixXd>&, const Map<VectorXd>&);
  };

  class SVD : public lm {
  public:
    SVD(const Map<MatrixXd>&, const Map<VectorXd>&);
  };

  class SymmEigen : public lm {
  public:
    SymmEigen(const Map<MatrixXd>&, const Map<VectorXd>&);
  };
}


#endif


src/big_fast_lm.cpp
// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; tab-width: 8 -*-
//
// fastLm.cpp: Rcpp/Eigen example of a simple lm() alternative
//
// Copyright (C) 2011 - 2015  Douglas Bates, Dirk Eddelbuettel and Romain Francois
//
// This file is part of RcppEigen.
//
// RcppEigen is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// RcppEigen is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// in file.path(R.home("share"), "licenses").  If not, see
// <http://www.gnu.org/licenses/>.

#include "big_fast_lm.h"
#include <bigmemory/MatrixAccessor.hpp>
#include <bigmemory/BigMatrix.h>
#if !defined(EIGEN_USE_MKL) // don't use R Lapack.h if MKL is enabled
#include <R_ext/Lapack.h>
#endif

namespace lmsol {
  using namespace Rcpp;
  using namespace RcppEigen;
  using Rcpp::_;
  using Rcpp::as;
  using Rcpp::CharacterVector;
  using Rcpp::clone;
  using Rcpp::List;
  using Rcpp::NumericMatrix;
  using Rcpp::NumericVector;
  using Rcpp::RObject;
  using Rcpp::wrap;

  using std::invalid_argument;
  using std::numeric_limits;

  lm::lm(const Map<MatrixXd> &X, const Map<VectorXd> &y)
    : m_X(X),
      m_y(y),
      m_n(X.rows()),
      m_p(X.cols()),
      m_coef(VectorXd::Constant(m_p, ::NA_REAL)),
      m_r(::NA_INTEGER),
      m_fitted(m_n),
      m_se(VectorXd::Constant(m_p, ::NA_REAL)),
      m_usePrescribedThreshold(false) {
  }

  lm& lm::setThreshold(const RealScalar& threshold) {
    m_usePrescribedThreshold = true;
    m_prescribedThreshold = threshold;
    return *this;
  }

  inline ArrayXd lm::Dplus(const ArrayXd& d) {
    ArrayXd   di(d.size());
    double  comp(d.maxCoeff() * threshold());
    for (int j = 0; j < d.size(); ++j) di[j] = (d[j] < comp) ? 0. : 1./d[j];
    m_r          = (di != 0.).count();
    return di;
  }

  MatrixXd lm::XtX() const {
    return MatrixXd(m_p, m_p).setZero().selfadjointView<Lower>().
    rankUpdate(m_X.adjoint());
  }

  /** Returns the threshold that will be used by certain methods such as rank().
  *
  *  The default value comes from experimenting (see "LU precision
  *  tuning" thread on the Eigen list) and turns out to be
  *  identical to Higham's formula used already in LDLt.
  *
  *  @return The user-prescribed threshold or the default.
  */
  RealScalar lm::threshold() const {
    return m_usePrescribedThreshold ? m_prescribedThreshold
    : numeric_limits<double>::epsilon() * m_p;
  }

  ColPivQR::ColPivQR(const Map<MatrixXd> &X, const Map<VectorXd> &y)
    : lm(X, y) {
    ColPivHouseholderQR<MatrixXd> PQR(X); // decompose the model matrix
    Permutation                  Pmat(PQR.colsPermutation());
    m_r                               = PQR.rank();
    if (m_r == m_p) {	// full rank case
      m_coef     = PQR.solve(y);
      m_fitted   = X * m_coef;
      m_se       = Pmat * PQR.matrixQR().topRows(m_p).
      triangularView<Upper>().solve(I_p()).rowwise().norm();
      return;
    }
    MatrixXd                     Rinv(PQR.matrixQR().topLeftCorner(m_r, m_r).
                                        triangularView<Upper>().
                                        solve(MatrixXd::Identity(m_r, m_r)));
    VectorXd                  effects(PQR.householderQ().adjoint() * y);
    m_coef.head(m_r)                  = Rinv * effects.head(m_r);
    m_coef                            = Pmat * m_coef;
    // create fitted values from effects
    // (can't use X*m_coef if X is rank-deficient)
    effects.tail(m_n - m_r).setZero();
    m_fitted                          = PQR.householderQ() * effects;
    m_se.head(m_r)                    = Rinv.rowwise().norm();
    m_se                              = Pmat * m_se;
  }

  QR::QR(const Map<MatrixXd> &X, const Map<VectorXd> &y) : lm(X, y) {
    HouseholderQR<MatrixXd> QR(X);
    m_coef                     = QR.solve(y);
    m_fitted                   = X * m_coef;
    m_se                       = QR.matrixQR().topRows(m_p).
    triangularView<Upper>().solve(I_p()).rowwise().norm();
  }


  Llt::Llt(const Map<MatrixXd> &X, const Map<VectorXd> &y) : lm(X, y) {
    LLT<MatrixXd>  Ch(XtX().selfadjointView<Lower>());
    m_coef            = Ch.solve(X.adjoint() * y);
    m_fitted          = X * m_coef;
    m_se              = Ch.matrixL().solve(I_p()).colwise().norm();
  }

  Ldlt::Ldlt(const Map<MatrixXd> &X, const Map<VectorXd> &y) : lm(X, y) {
    LDLT<MatrixXd> Ch(XtX().selfadjointView<Lower>());
    Dplus(Ch.vectorD());	// to set the rank
    //FIXME: Check on the permutation in the LDLT and incorporate it in
    //the coefficients and the standard error computation.
    //	m_coef            = Ch.matrixL().adjoint().
    //	    solve(Dplus(D) * Ch.matrixL().solve(X.adjoint() * y));
    m_coef            = Ch.solve(X.adjoint() * y);
    m_fitted          = X * m_coef;
    m_se              = Ch.solve(I_p()).diagonal().array().sqrt();
  }

  SVD::SVD(const Map<MatrixXd> &X, const Map<VectorXd> &y) : lm(X, y) {
    JacobiSVD<MatrixXd>  UDV(X.jacobiSvd(ComputeThinU|ComputeThinV));
    MatrixXd             VDi(UDV.matrixV() *
      Dplus(UDV.singularValues().array()).matrix().asDiagonal());
    m_coef                   = VDi * UDV.matrixU().adjoint() * y;
    m_fitted                 = X * m_coef;
    m_se                     = VDi.rowwise().norm();
  }

  SymmEigen::SymmEigen(const Map<MatrixXd> &X, const Map<VectorXd> &y)
    : lm(X, y) {
    SelfAdjointEigenSolver<MatrixXd> eig(XtX().selfadjointView<Lower>());
    MatrixXd   VDi(eig.eigenvectors() *
      Dplus(eig.eigenvalues().array()).sqrt().matrix().asDiagonal());
    m_coef         = VDi * VDi.adjoint() * X.adjoint() * y;
    m_fitted       = X * m_coef;
    m_se           = VDi.rowwise().norm();
  }

  enum {LLT_t = 0, LDLT_t}; //, SymmEigen_t, QR_t};

  static inline lm do_lm(const Map<MatrixXd> &X, const Map<VectorXd> &y, int type) {
    switch(type) {
    case LLT_t:
      return Llt(X, y);
    case LDLT_t:
      return Ldlt(X, y);
    }
    throw invalid_argument("invalid type");
    return Ldlt(X, y);	// -Wall
  }

  List fastLm(XPtr<BigMatrix> Xs, Rcpp::NumericVector ys, int type)
  {

    XPtr<BigMatrix> bMPtr(Xs);

    unsigned int typedata = bMPtr->matrix_type();

    if (typedata != 8)
    {
      throw Rcpp::exception("type for provided big.matrix not available");
    }

    const Map<MatrixXd>  X = Map<MatrixXd>((double *)bMPtr->matrix(), bMPtr->nrow(), bMPtr->ncol()  );
    const Map<VectorXd>  y(as<Map<VectorXd> >(ys));
    Index                n = X.rows();
    if ((Index)y.size() != n) throw invalid_argument("size mismatch");


    // Select and apply the least squares method
    lm                 ans(do_lm(X, y, type));

    // Copy coefficients and install names, if any
    NumericVector     coef(wrap(ans.coef()));

    List          dimnames( wrap(bMPtr->column_names()) );
    if (dimnames.size() > 1) {
      RObject   colnames = dimnames[1];
      if (!(colnames).isNULL())
        coef.attr("names") = clone(CharacterVector(colnames));
    }

    VectorXd         resid = y - ans.fitted();
    int               rank = ans.rank();
    int                 df = (rank == ::NA_INTEGER) ? n - X.cols() : n - rank;
    double               s = resid.norm() / std::sqrt(double(df));
    // Create the standard errors
    VectorXd            se = s * ans.se();

    return List::create(_["coefficients"]  = coef,
                        _["se"]            = se,
                        _["rank"]          = rank,
                        _["df.residual"]   = df,
                        _["residuals"]     = resid,
                        _["s"]             = s,
                        _["fitted.values"] = ans.fitted());
  }
}

// This defines the R-callable function 'bigLm'
RcppExport SEXP bigLm_Impl(SEXP X, SEXP y, SEXP type)
{
  BEGIN_RCPP
  Rcpp::RObject __result;
  Rcpp::RNGScope __rngScope;
  Rcpp::traits::input_parameter< Rcpp::XPtr<BigMatrix> >::type X_(X);
  Rcpp::traits::input_parameter< Rcpp::NumericVector >::type y_(y);
  Rcpp::traits::input_parameter< int >::type type_(type);
  __result = Rcpp::wrap(lmsol::fastLm(X_, y_, type_));
  return __result;
  END_RCPP
}


R/fast_big_lm.R
## bigLm.R: Rcpp/Eigen/bigmemory implementation of lm()
##
## Copyright (C)  2011 - 2015  Douglas Bates, Dirk Eddelbuettel and Romain Francois
##
## Copyright (C)  2016  Jared Huling
##
## This file is based off of code from RcppEigen.
##
## RcppEigen is free software: you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 2 of the License, or
## (at your option) any later version.
##
## RcppEigen is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with RcppEigen.  If not, see <http://www.gnu.org/licenses/>.



#' fast and memory efficient linear model fitting
#'
#' @param X input model matrix. must be a big.matrix object (type = 8 for double)
#' @param y numeric response vector of length nobs.
#' @param method an integer scalar with value 0 for the LLT Cholesky or 1 for the LDLT Cholesky
#' @return A list with the elements
#' \item{coefficients}{a vector of coefficients}
#' \item{se}{a vector of the standard errors of the coefficient estimates}
#' \item{rank}{a scalar denoting the computed rank of the model matrix}
#' \item{df.residual}{a scalar denoting the degrees of freedom in the model}
#' \item{residuals}{the vector of residuals}
#' \item{s}{a numeric scalar - the root mean square for residuals}
#' \item{fitted.values}{the vector of fitted values}
#' @export
#' @examples
#'
#' library(bigmemory)
#'
#' nrows <- 50000
#' ncols <- 50
#' bkFile <- "bigmat2.bk"
#' descFile <- "bigmatk2.desc"
#' bigmat <- filebacked.big.matrix(nrow=nrows, ncol=ncols, type="double",
#'                                 backingfile=bkFile, backingpath=".",
#'                                 descriptorfile=descFile,
#'                                 dimnames=c(NULL,NULL))
#'
#' # Each column value with be the column number multiplied by
#' # samples from a standard normal distribution.
#' set.seed(123)
#' for (i in 1:ncols) bigmat[,i] = rnorm(nrows)*i
#'
#' y <- rnorm(nrows) + bigmat[,1]
#'
#' system.time(lmr1 <- bigLmPure(bigmat, y))
#'
#' system.time(lmr2 <- lm.fit(x = bigmat[,], y = y))
#'
#' max(abs(coef(lmr1) - coef(lmr2)))
#'
#'
bigLmPure <- function(X, y, method = 0L) {


    method <- as.integer(method[1])

    if (method > 1L)
    {
        stop("invalid method")
    }

    cnames <- colnames(X)
    if (is.null(cnames)) cnames <- paste0("X", 1:ncol(X))

    stopifnot( is.big.matrix(X), is.numeric(y), NROW(y) == nrow(X) )

    res <- .Call("bigLm_Impl", X@address, y, method, cnames, PACKAGE="bigFastlm")
    names(res$coefficients) <- cnames
    res
}

#' fast and memory efficient linear model fitting
#'
#' @param X input model matrix. must be a big.matrix object (type = 8 for double)
#' @param y numeric response vector of length nobs.
#' @param method an integer scalar with value 0 for the LLT Cholesky or 1 for the LDLT Cholesky
#' @return A list object with S3 class "bigLm" with the elements
#' \item{coefficients}{a vector of coefficients}
#' \item{se}{a vector of the standard errors of the coefficient estimates}
#' \item{rank}{a scalar denoting the computed rank of the model matrix}
#' \item{df.residual}{a scalar denoting the degrees of freedom in the model}
#' \item{residuals}{the vector of residuals}
#' \item{s}{a numeric scalar - the root mean square for residuals}
#' \item{fitted.values}{the vector of fitted values}
#' @useDynLib bigFastlm
#' @import Rcpp
#' @export
#' @examples
#'
#' library(bigmemory)
#'
#' nrows <- 50000
#' ncols <- 50
#' bkFile <- "bigmat.bk"
#' descFile <- "bigmatk.desc"
#' bigmat <- filebacked.big.matrix(nrow=nrows, ncol=ncols, type="double",
#'                                 backingfile=bkFile, backingpath=".",
#'                                 descriptorfile=descFile,
#'                                 dimnames=c(NULL,NULL))
#'
#' # Each column value with be the column number multiplied by
#' # samples from a standard normal distribution.
#' set.seed(123)
#' for (i in 1:ncols) bigmat[,i] = rnorm(nrows)*i
#'
#' y <- rnorm(nrows) + bigmat[,1]
#'
#' system.time(lmr1 <- bigLm(bigmat, y))
#'
#' system.time(lmr2 <- lm.fit(x = bigmat[,], y = y))
#'
#' max(abs(coef(lmr1) - coef(lmr2)))
#'
#'
bigLm <- function(X, ...) UseMethod("bigLm")



#' summary method for bigLm fitted objects
#'
#' @param object bigLm fitted object
#' @param ... not used
#' @return a summary.bigLm object
#' @rdname summary
#' @method summary bigLm
#' @export
#' @examples
#'
#' library(bigmemory)
#'
#' nrows <- 50000
#' ncols <- 15
#' bkFile <- "bigmat4.bk"
#' descFile <- "bigmatk4.desc"
#' bigmat <- filebacked.big.matrix(nrow=nrows, ncol=ncols, type="double",
#'                                 backingfile=bkFile, backingpath=".",
#'                                 descriptorfile=descFile,
#'                                 dimnames=c(NULL,NULL))
#'
#' # Each column value with be the column number multiplied by
#' # samples from a standard normal distribution.
#' set.seed(123)
#' for (i in 1:ncols) bigmat[,i] = rnorm(nrows)*i
#'
#' y <- rnorm(nrows) + bigmat[,1] - bigmat[,2]
#'
#' system.time(lmr1 <- bigLm(bigmat, y))
#'
#' summary(lmr1)
#'
#'
summary.bigLm <- function(object, ...)
{
    coef <- object$coefficients
    se   <- object$se
    tval <- coef/se

    object$coefficients <- cbind(Estimate     = coef,
                                 "Std. Error" = se,
                                 "t value"    = tval,
                                 "Pr(>|t|)"   = 2*pt(-abs(tval), df=object$df))

    ## cf src/stats/R/lm.R and case with no weights and an intercept
    f <- object$fitted.values
    r <- object$residuals
    #mss <- sum((f - mean(f))^2)
    mss <- if (object$intercept) sum((f - mean(f))^2) else sum(f^2)
    rss <- sum(r^2)

    object$r.squared <- mss/(mss + rss)
    df.int <- if (object$intercept) 1L else 0L
    n <- length(f)
    rdf <- object$df
    object$adj.r.squared <- 1 - (1 - object$r.squared) * ((n - df.int)/rdf)
    class(object) <- "summary.bigLm"
    object
}


#' bigLm default
#'
#' @param ... not used
#' @rdname bigLm
#' @method bigLm default
#' @export
bigLm.default <- function(X, y, method = 0L, ...) {

    y <- as.numeric(y)

    res <- bigLmPure(X, y, method)
    res$call <- match.call()
    res$intercept <- any(big.colMax(X) == big.colMin(X))

    class(res) <- "bigLm"
    res
}

#' print method for bigLm objects
#'
#' @rdname print
#' @method print bigLm
#' @export
print.bigLm <- function(x, ...) {
    cat("\nCall:\n")
    print(x$call)
    cat("\nCoefficients:\n")
    print(x$coefficients, digits=5)
}

#' print method for summary.bigLm objects
#'
#' @param x a "summary.bigLm" object
#' @param ... not used
#' @rdname print
#' @method print summary.bigLm
#' @export
print.summary.bigLm <- function(x, ...) {
    cat("\nCall:\n")
    print(x$call)
    cat("\nResiduals:\n")
    digits <- max(3, getOption("digits") - 3)
    print(summary(x$residuals, digits=digits)[-4])
    cat("\n")

    printCoefmat(x$coefficients, P.values=TRUE, has.Pvalue=TRUE, ...)
    cat("\nResidual standard error: ", formatC(x$s, digits=digits), " on ",
        formatC(x$df), " degrees of freedom\n", sep="")
    cat("Multiple R-squared: ", formatC(x$r.squared, digits=digits),
        ",\tAdjusted R-squared: ",formatC(x$adj.r.squared, digits=digits),
        "\n", sep="")
    invisible(x)
}

#' Prediction method for bigLm fitted objects
#'
#' @param object fitted "bigLm" model object
#' @param newdata big.matrix object. If NULL, then fitted values are returned
#' @param ... not used
#' @return An object depending on the type argument
#' @importFrom bigmemory is.big.matrix as.big.matrix
#' @importFrom stats coef fitted pt printCoefmat
#' @export
#' @examples
#'
#' library(bigmemory)
#'
#' nrows <- 50000
#' ncols <- 50
#' bkFile <- "bigmat3.bk"
#' descFile <- "bigmatk3.desc"
#' bigmat <- filebacked.big.matrix(nrow=nrows, ncol=ncols, type="double",
#'                                 backingfile=bkFile, backingpath=".",
#'                                 descriptorfile=descFile,
#'                                 dimnames=c(NULL,NULL))
#'
#' # Each column value with be the column number multiplied by
#' # samples from a standard normal distribution.
#' set.seed(123)
#' for (i in 1:ncols) bigmat[,i] = rnorm(nrows)*i
#'
#' y <- rnorm(nrows) + bigmat[,1]
#'
#' system.time(lmr1 <- bigLm(bigmat, y))
#'
#'
#' preds <- predict(lmr1, newdata = bigmat)
#'
predict.bigLm <- function(object, newdata = NULL, ...) {
    if (is.null(newdata)) {
        y <- fitted(object)
    } else {
        stopifnot(is.big.matrix(newdata))
        y <- as.vector(newdata %*% coef(object))
    }
    y
}
